src/ressa/explorer.rs

```diff
@@ -1,118 +1,59 @@
 /// WARNING: HERE THERE BE MACROS
-use super::{ressa_node_parse, NodePattern, ParserContext};
+use super::{ressa_node_parse, IntoRessaNode, NodePattern, NodePatternParser, ParserContext};
+use super::{Indexable, LaastIndex};
 use crate::ast::*;
 use crate::prophet::*;
 use enum_dispatch::enum_dispatch;
 
-#[enum_dispatch]
-pub enum ProphetNode {
-    ClassOrInterfaceComponent(ClassOrInterfaceComponent),
-    MethodComponent(MethodComponent),
-    MethodParamComponent(MethodParamComponent),
-    FieldComponent(FieldComponent),
-    AnnotationComponent(AnnotationComponent),
-    AnnotationValuePair(AnnotationValuePair),
-}
-
 /// Describes how to visit the nodes in the AST to find nodes of interest
 #[enum_dispatch(Node)]
 #[enum_dispatch(Expr)]
 #[enum_dispatch(Stmt)]
-#[enum_dispatch(ProphetNode)]
 pub trait RessaNodeExplorer {
-    fn explore(&self, pattern: &mut NodePattern, ctx: &mut ParserContext) -> Option<()>;
+    fn explore(
+        &self,
+        pattern: &mut NodePattern,
+        ctx: &mut ParserContext,
+        index: &LaastIndex,
+    ) -> Option<()>;
 }
 
-/// Create a no-operation implementation of exploring a node
-#[macro_export]
-macro_rules! ressa_dispatch_default_impl {
+/// Generate a default explore implementation that delegates to the child fields (blanket implementation caused problems)
```

Comment:
What was the issue with the blanket implementation?

Reply:
The issue is that we need two different implementations of `explore` for two different subsets--those we match on, and those we just recurse on. By adding a blanket implementation for one, you can't implement the other. I heard about `impl specialization` to potentially handle cases like this, but it doesn't seem to work for me when I tried it.

Reply:
Looks like specialization still unstable so probably best to not use it https://rust-lang.github.io/rfcs/1210-impl-specialization.html